### ISSUE TITLE
dependency injection between controller and data provider

### ISSUE THREAD

**Issue by shanlin2dltk** [jonataslaw/getx#1535]
**ATTENTION: DO NOT USE THIS FIELD TO ASK SUPPORT QUESTIONS. USE THE PLATFORM CHANNELS FOR THIS. THIS SPACE IS DEDICATED ONLY FOR FEATURE REQUESTS**

**Is your feature request related to a problem? Please describe.**
Since dependency injection between view and controller is well supported in GetX, it is pretty natural to ask whether dependency injection between controller and data layer, or controller to controller can be done.  In a complex, real project, the need to have these kind of dependency injection is obvious.  If this feature is already supported, please give examples on how this can be done.

**Comment by ahmednfwela** [comment]
it is supported, using the same concepts of view-controller dependency injection

simply call `Get.find<T>` and it will resolve the type for you.
doesn't matter if you call it from a view, or from another controller.

**Comment by shanlin2dltk** [comment]
Thanks @ahmednfwela for pointing it out.  It is better if a short example can be put in the documentation.  Another question is: since dependency injection is so great for unit testing as well, is there any example or documentation on how to unit testing controllers and views?

**Comment by raaone7** [comment]
@ahmednfwela , this is what i tried, with no luck. 


eg. 
- login view - took username and password, and called logincontroller.login
- logincontroller.login called provider , and created the loginuser model (nested class)
- logincontroller.login stored the loginuser model into getx 
_note : loginuser model is not observable here , and how to do that ?_ 
- logincontroller.login passed result to login view
- login view routed to home view 
- home view had binding of login user - perfect till this poing
- tried to access loginuser model home view - raw works
- tried to change loginuser model in home controller as results of any action - 
- home view - doesnot react, and cant use Obx with login user, since login user is not reactive and is stored inside getx

some code

login controller code to store the model object in GetX
```
      // set the loginUser in the getx so that it can be used in views
      Get.putAsync<LoginUser>(() async {
        return loginUser;
      }, permanent: true);

```

home view retrieval 
```
  final LoginUser loginUser = Get.find();
```

note in both the above steps, loginUser is not reactive ? how to make it reactive and use it with Obx

**Comment by raaone7** [comment]
@shanlin2dltk - you got a fair point, i can't find any working example in google. i understand getX is super powerful, but lack of documentation and examples make it really hard to troubleshoot issues

**Comment by danbo** [comment]
raaone7,

From what I can tell (since you haven't shared the details / code for loginUser), it is a complex object containing a field called name and you are trying to make the complex object observable, changing a field and expecting Obx to pick up that change.

If you look at the [documentation](https://github.com/jonataslaw/getx/blob/master/documentation/en_US/state_management.md#reactive-state-manager), you will see that observable types are actually Rx**<something>**, whether you declare them yourself or use the **.obs** extension. You will notice that there is no Rx<> type for a complex object, only primitives, lists and maps.

This means that you need to make each field such as name, observable within your loginUser. ie

```
import 'package:get/get.dart';
class loginUser {
  final Rx<String> name = new Rx<String>(""); // OR
  final name = "".obs;
}
```

then somewhere register an instance of loginUser

```
Get.putAsync<LoginUser>(() async {
        return loginUser;
      }, permanent: true);
}
```

then in some view, make it available and use it.

```
class HomeScreen extends StatelessWidget {
  late final LoginUser _loginUser;

  HomeScreen({Key? key}) : super(key: key) {
    _loginUser = Get.find<LoginUser>();
  }

  void _handleButtonClicked() {
    _loginUser.name.value = "changed! " + DateTime.now().toLocal().toString();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(),
      body: Center(
        child: Column(children: [
          Obx(() => Text("${_loginUser.name}")),
          ElevatedButton(
            onPressed: _handleButtonClicked,
            child: Text("Press Me"),
          )
        ]),
      ),
    );
  }
```

This is something that's a bit easier to figure out if you've dealt with UI state management in the past, ie in .NET WPF or React/Redux. Though the concepts of reactive programming have been around for a while, the [original Rx](https://en.wikipedia.org/wiki/ReactiveX) implementation was pioneered a decade ago in .NET and was also popular in the UI space as an alternative to the UI specific notification mechanisms that were more expensive and less flexible. It has since been ported to Java, JavaScript among other languages for UI and non UI use cases. 

React, like WPF has its own specific UI notification mechanisms which are also fine for simpler use cases but not nearly as powerful as a more reactive based approach - enter Redux. With Redux however, the strategy is to lift state out of things like complex objects and controllers and have a global state store which acts as one global observable view model with the intention of staying as flat and as far away from complex types as possible, and this is where the idea of shedding complex observable types comes from. It should be noted that even with this architecture, it is still easy to have too much listening or broad targeting which results in unnecessary re-renders and many cases the Rx approach is still preferred due to the granularity and reactive tuning you can do on messages.

So in summary, yes, the documentation could and should be improved (especially to be more concise with more concrete examples), however, to properly explain everything would go beyond the scope of GetX itself and more into architecture patterns for (UI) development which span multiple concepts and technologies.

**Comment by shanlin2dltk** [comment]
@danbo according to the documentation you cited, LoginUser can be observable.  In its example, it said:
`// Custom classes - it can be any class, literally
final user = User().obs;`

**Comment by danbo** [comment]
> 
> 
> @danbo according to the documentation you cited, LoginUser can be observable. In its example, it said:
> `// Custom classes - it can be any class, literally final user = User().obs;`

@shanlin2dltk, I noticed that as well and you beat me to it. I have updated my suggestion.
